Added job conf section to ansible-galaxy tutorial (#1991)
galaxyproject/training-material: topics/admin/tutorials/ansible-galaxy/tutorial.md

    tutorials:
      - ansible
  - type: "none"

>     title: "A VM with at least 2 vCPUs and 4 GB RAM with Ubuntu 18.04 - 20.04 for preference."

Reviewer: ```suggestion
    title: "A VM with at least 2 vCPUs and 4 GB RAM, preferably running Ubuntu 18.04 - 20.04."
```

## Job Configuration

One of the most important configuration files for a large Galaxy server is the `job_conf.xml` file. This file tells Galaxy where to run all of the jobs that users execute. If Galaxy can't find a job conf file or none have been specified in the `galaxy.yml` file, it will use a default configuration. The default configuration is the `job_conf.xml.sample_basic` file. This file is located at: `/srv/galaxy/server/lib/galaxy/config/sample/job_conf.xml.sample_basic`, though there is a symlink to the file in `/srv/galaxy/server/config`.

The job configuration file allows Galaxy to run jobs in multiple locations using a variety of different mechanisms. Some of these mechanisms include:
* Local - Galaxy runs jobs on the same computer that Galaxy itself is running on.
* DRMAA - Galaxy can connect to a cluster and submit jobs via a distributed resource manager such as Slurm, Condor, PBS Torque or Sun Grid Engine.
* Pulsar - Galaxy can also send jobs to remote compute hosts over the internet using Pulsar.

### The job conf file - basics

> The `job_conf.xml` has three basic sections:

Reviewer: ```suggestion
The `job_conf.xml` file has three basic sections:
```

> * **Plugins** - This section lists the types of resources that this Galaxy server is configured to use and tells Galaxy to load the drivers for each type.

Reviewer: ```suggestion
* **Plugins** - This section lists the types of job management systems that this Galaxy server is configured to use and tells Galaxy to load the drivers for each type.
```

> * **Destinations** - This section lists the different locations or queues etc that Galaxy can send jobs to. Each one has a name and uses a *plugin* to communicate with that location. They can specify things like the number of cpu's, amount of RAM to be allocated etc for DRMAA locations. Usually one of the destinations is set to be the default.

Reviewer: ```suggestion
* **Destinations** - This section lists the different locations, queues, etc. that Galaxy can send jobs to. Each one has a name and uses a *plugin* to communicate with that location. They can specify things like the number of CPUs, amount of RAM to be allocated, etc. for DRMAA locations. Usually, one of the destinations is set to be the default.
```

> * **Tools** - This section lists the various tools that Galaxy would like to send to the non-default *destination*. Each line in this section pairs up a tool in Galaxy with a particular job *destination*. Then everytime Galaxy gets a job for that particular tool, it is always sent to that *destination*.

Reviewer: ```suggestion
* **Tools** - This section lists the various tools that you would like to send to a non-default *destination*. Each line in this section pairs up a tool in Galaxy with a particular job *destination*. Every time Galaxy gets a job for that particular tool, it is always sent to that *destination*.
```

The basic `job_conf.xml` file looks like this:

{% raw %}
```xml
<job_conf>
    <plugins workers="4">
        <plugin id="local" type="runner" load="galaxy.jobs.runners.local:LocalJobRunner"/>
    </plugins>
    <destinations>

>         <destination id="local" runner="local"/>

Reviewer: In Barcelona, as we added to the `job_conf.xml` there was some confusion over the distinction between destination IDs and runner IDs. This is copied straight from `job_conf.xml.sample_basic` but I wonder if we should change this to:

```xml
        <plugin id="local_plugin" type="runner" load="galaxy.jobs.runners.local:LocalJobRunner"/>
```

and:

```xml
        <destination id="local_destination" runner="local_plugin"/>
```

    </destinations>
    <tools>
    </tools>
</job_conf>
```

The above `job_conf.xml` file defines a *plugin* and *destination* to allow Galaxy to run user jobs on the local computer (i.e. The computer that Galaxy is running on.)

Firstly, the plugins section contains a plugin called "local" which is of type "runner" and then loads the python code module for supporting local jobs. Next the destinations section contains a destination called "local" using the runner "local". As this is the only destination specified, it is also the default. So now everytime a user clicks "Execute" on a tool form, Galaxy will run the corresponding job locally using the python code specified.

There are a lot of other plugins available for Galaxy for using other resources such as docker containers, kubernetes hosts, Pulsar destinations and HPC clusters to name a few. See the Galaxy documentation on [job configuration](https://docs.galaxyproject.org/en/master/admin/jobs.html) for more details on these plugins and their configuration. There is also an advanced sample job conf file located at: `/srv/galaxy/server/lib/galaxy/config/sample/job_conf.xml.sample_advanced`

> {% endraw %}

Reviewer: Should this be moved up after the `job_conf.xml` block?

>
>    {% raw %}
>    ```yaml
>    galaxy_config:
>      galaxy:
>        # ... existing configuration options in the `galaxy` section ...
>        job_config_file: "{{ galaxy_config_dir }}/job_conf.xml"
>    ```
>    {% endraw %}
>
>    And then deploy the new config file using the `galaxy_config_templates` var in your group vars:
>
>    {% raw %}
>    ```yaml
>    galaxy_config_templates:
>      # ... possible existing config file definitions
>      - src: templates/galaxy/config/job_conf.xml.j2
>        dest: "{{ galaxy_config.galaxy.job_config_file }}"
>    ```
>    {% endraw %}
>
> 4. Run the playbook. At the very end, you should see output like the following indicating that Galaxy has been restarted:
>
>    ```
>    RUNNING HANDLER [restart galaxy] ****************************************
>    changed: [galaxy.example.org]
>    ```
> 5. Checkout the new job_conf.xml file.
>
>    > ### {% icon code-in %} Input: Bash
>    > ```bash
>    > cat /srv/galaxy/config/job_conf.xml
>    > ```
>    {: .code-in}
>
>    > ### {% icon code-out %} Output: Bash

> >    > ```ini

Reviewer: ```suggestion
>    > ```xml
```

In order to be the administrator user, you will need to register an account with the same email address you used in the group variables under the `admin_users` setting.

> ## Job Configuration (Optional)

Reviewer 2: ```suggestion
## Job Configuration
```

I kinda think it should be mandatory, it's pretty quick to do this step, and helps get them on the right path for one of the first questions admins have "how can I connect it to my cluster"

There are a lot of other plugins available for Galaxy for using other resources such as docker containers, kubernetes hosts, Pulsar destinations and HPC clusters to name a few. See the Galaxy documentation on [job configuration](https://docs.galaxyproject.org/en/master/admin/jobs.html) for more details on these plugins and their configuration. There is also an advanced sample job conf file located at: `/srv/galaxy/server/lib/galaxy/config/sample/job_conf.xml.sample_advanced`

{% endraw %}

> ### {% icon hands_on %} Hands-on: Job Conf
>
> 1. If the folder does not exist, create `templates/galaxy/config` next to your `galaxy.yml` playbook (`mkdir -p templates/galaxy/config/`).
>
> 2. Create `templates/galaxy/config/job_conf.xml.j2` with the following contents:
>
>    ```xml
>    <job_conf>
>        <plugins workers="4">
>            <plugin id="local" type="runner" load="galaxy.jobs.runners.local:LocalJobRunner"/>
>        </plugins>
>        <destinations default="local">
>            <destination id="local" runner="local"/>
>        </destinations>
>        <tools>
>        </tools>
>    </job_conf>
>    ```
>

> > 3. Inform `galaxyproject.galaxy` of where you would like the `job_conf.xml` to reside in your group variables:

Reviewer 2: ```suggestion
> 3. Inform `galaxyproject.galaxy` of where you would like the `job_conf.xml` to reside, by setting it in your `group_vars/galaxyservers.yml`:
```

got this comment somewhere else that "it was too unclear" and ok, it should really be in their skill set to figure out if they're comprehension is good, but, maybe we'll give them a helping hand in the first session.

>
>    {% raw %}
>    ```yaml
>    galaxy_config:
>      galaxy:
>        # ... existing configuration options in the `galaxy` section ...
>        job_config_file: "{{ galaxy_config_dir }}/job_conf.xml"
>    ```
>    {% endraw %}
>
>    And then deploy the new config file using the `galaxy_config_templates` var in your group vars:
>
>    {% raw %}
>    ```yaml
>    galaxy_config_templates:
>      # ... possible existing config file definitions
>      - src: templates/galaxy/config/job_conf.xml.j2
>        dest: "{{ galaxy_config.galaxy.job_config_file }}"
>    ```
>    {% endraw %}
>
> 4. Run the playbook. At the very end, you should see output like the following indicating that Galaxy has been restarted:
>
>    ```
>    RUNNING HANDLER [restart galaxy] ****************************************
>    changed: [galaxy.example.org]
>    ```

> > 5. Checkout the new job_conf.xml file in `/srv/galaxy/config`.

Reviewer 2: could use one of the IO blocks here to show `cat /srv/galaxy/config/job_conf` and showing the output? Dunno if that's more instructive, but the green box might act as a better call to action than a bullet point.

Author: Good thinking. I'll add that in
